core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageDuplicator.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import org.reactivestreams.Subscriber;
+
+/**
+ * A duplicator that duplicates a {@link StreamMessage} into one or more {@link StreamMessage}s,
+ * which publish the same elements.
+ *
+ * <p>Only one subscriber can subscribe to a {@link StreamMessage}. If you want to subscribe to it
+ * multiple times, use {@link StreamMessageDuplicator} which is returned by calling
+ * {@link StreamMessage#toDuplicator()}.
+ * <pre>{@code
+ * StreamMessage<String> streamMessage = ...
+ * StreamMessageDuplicator<String> duplicator = streamMessage.toDuplicator();
+ * // streamMessage.subscribe(...) will throw an exception. You cannot subscribe to streamMessage anymore.
+ *
+ * // Duplicate the stream as many as you want to subscribe.
+ * StreamMessage<String> duplicatedStreamMessage1 = duplicator.duplicate();
+ * StreamMessage<String> duplicatedStreamMessage2 = duplicator.duplicate();
+ * duplicatedStreamMessage1.subscribe(...);
+ * duplicatedStreamMessage2.subscribe(...);
+ *
+ * duplicator.close(); // You should call close to clean up the resources.
+ * }</pre>
+ *
+ * <p>If you subscribe to the {@linkplain #duplicate() duplicated stream message} with the
+ * {@link SubscriptionOption#WITH_POOLED_OBJECTS}, the published elements can be shared across
+ * {@link Subscriber}s. So do not manipulate the data unless you copy them.
+ *
+ * <p>To clean up the resources, you have to call one of {@linkplain #duplicate(boolean) duplicate(true)},
+ * {@link #close()} or {@link #abort()}. Otherwise, memory leak might happen.</p>
+ *
+ * @param <T> the type of elements
+ */
+public interface StreamMessageDuplicator<T> {
+
+    /**
+     * Returns a new {@link StreamMessage} that publishes the same elements with the {@link StreamMessage}
+     * that this duplicator is created from.
+     *
+     * @param lastStream whether to prevent further duplication
+     */
+    StreamMessage<T> duplicate(boolean lastStream);
```

Comment:
Do we need this? From what I understand from the javadocs, these two are equivalent

```java
duplicator.duplicate(true);
duplicator.close();
```

```java
duplicator.duplicate();
duplicator.close();
```

Reply:
Yes, they are equivalent.
Actually, we don't have to call `close` in the first example.
```java
duplicator.duplicate(true).subscribe(...);
// Equivalent to 
duplicator.duplicate().subscribe(...);
duplicator.close();
```

We need the first code for the response in `RetryingClient`. We have to return the last `HttpResonse` to the user API, so when the last response is subscribed and complete, the resources are cleaned up. (We don't know when we need to call `close()` because we don't know the duplicate `HttpResponse` is subscribed by the user or not.)

We need the second one for the request. The duplicated `HttpRequest` is already subscribed by the `HttpRequestSubscriber` and we don't want to duplicate anymore so just call `close()`.

Reply:
But `close` javadoc says that duplicated streams are not cleaned up until they finish publication. So shouldn't it be fine to replace any instance of `duplicate(true)` with `duplicate()` and `close()`, even for `RetryingClient`?

The class-level javadoc recommends always calling `close` to make sure resources are cleaned up, which makes sense, but in that case, if `duplicate(true)` is equivalent, I think we should remove it if it's redundant.

Reply:
> duplicate(true) with duplicate() and close(), even for RetryingClient response?

The problem is this throws an exception:
```java
StreamMessage message = duplicator.duplicate();
duplicator.close();
message.subscribe(...) // This throw an exception because the duplicator is closed already.
```
So I have to change that a subscriber is able to subscribe to the duplicated stream which is created before `duplicator.close()` is called.

Let me see what I can do. 🤔

Reply:
If it's possible to support that, I think that could be cleanest, and it seems doable - it's normal for a stream that's not subscribed or aborted to be leaked, so it's not necessary for `close` to be cleaning up unsubscribed streams. 

But otherwise, now I understand the example better - so if we can't fix that behavior, the javadoc for close needs to be extended to mention that any streams that haven't been subscribed to will be cleaned up right away, which isn't the impression I got from the current javadoc.

Reply:
> If it's possible to support that, I think that could be cleanest, and it seems doable

Yes, it's much cleaner and I think I need to fix that. 😄 Thanks!

Reply:
@anuraaga Changed the logic as you suggested. PTAL 🙇

Reply:
I guess some users might want to release the duplicator automatically. How about extending `SafeCloseable`?
```java
public interface HttpRequestDuplicator extends StreamMessageDuplicator<HttpObject>, SafeCloseable {
    ...
    @Override
    void close();
}
```

Reply:
That's a good idea! Fixed. Thanks!